File: 56.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */

bool judge(vector<Interval>& intervals, vector<Interval>::iterator i) {
	bool res = false;
	for(vector<Interval>::iterator j = i + 1; j != intervals.end(); ) {
		//if(i == j) continue;
		if(i->start <= j->end && i->start >= j->start) {
			i->start = j->start;
			i->end = i->end > j->end ? i->end : j->end;
			j = intervals.erase(j);
			res = true;
		}else if(j->start <= i->end && j->start >= i->start) {
			//i->start = j->start;
			i->end = i->end > j->end ? i->end : j->end;
			j = intervals.erase(j);
			res = true;
		}else {
			++j;
		}
	}
	return res;
}

class Solution {
public:
    vector<Interval> merge(vector<Interval>& intervals) {
  		vector<Interval> res;
  		//for(int i = 0; i < intervals.size(); ++i) {
  		for(auto i = intervals.begin(); i != intervals.end(); ++i) {
  			while(judge(intervals, i));
  		}	
      	return intervals;
    }
};

File: 56.py
from typing import List


class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if not intervals:
            return []
        intervals.sort()
        ans = [intervals[0][:]]
        idx = 0
        for i in range(1, len(intervals)):
            if intervals[i][0] <= ans[idx][1]:
                ans[idx][1] = max(intervals[i][1], ans[idx][1])
            else:
                ans.append(intervals[i][:])
                idx += 1
        return ans

if __name__ == "__main__":
    s = Solution()
    l = [[1,7],[2,6],[8,10],[15,18]]
    # l = [[1,4],[4,5]]
    # l = []
    print(s.merge(l))